File: Scalemon.MassaKInterop/ScaleDriver100.vb

Public Class ScaleDriver100
    Implements IScaleDriver

    Private _scale As MassaKDriver100.Scales
    Private _response As Long


    Public Sub New()
        _scale = New MassaKDriver100.Scales()
        _response = 0
    End Sub

    Public Property PortConnection As String Implements IScaleDriver.PortConnection
        Get
            Return _scale.Connection
        End Get
        Set(value As String)
            _scale.Connection = value
        End Set
    End Property

    Public ReadOnly Property Weight As Decimal Implements IScaleDriver.Weight
        Get
            Return CDec(_scale.Weight) / 100D
        End Get
    End Property

    Public ReadOnly Property Stable As Boolean Implements IScaleDriver.Stable
        Get
            Return _scale.Stable = 1
        End Get
    End Property

    Public ReadOnly Property LastResponseText As String Implements IScaleDriver.LastResponseText
        Get
            Dim _message As String
            Select Case _response
                Case 0
                    _message = "Ошибок нет"
                Case 1
                    _message = "Связь с весами не установлена"
                Case 2
                    _message = "Ошибка обмена данных с весами"
                Case 3
                    _message = "Весы не готовы к передаче данных"
                Case 4
                    _message = "Параметр не поддерживается весами"
                Case 5
                    _message = "Установка параметра невозможна"
                Case 7
                    _message = "Невозможно выполнить команду или команда не поддерживается"
                Case 8
                    _message = "Нагрузка на весовом устройстве превышает НПВ"
                Case 9
                    _message = "Весовое устройство не в режиме взвешивания"
                Case 10
                    _message = "Ошибка входных данных"
                Case 11
                    _message = "Ошибка сохранения данных"
                Case 16
                    _message = "Интерфейс Wi-Fi не поддерживается"
                Case 17
                    _message = "Интерфейс Ethernet не поддерживается"
                Case 21
                    _message = "Установка нуля невозможна из-за наличия нагрузки на платформе"
                Case 23
                    _message = "Нет связи с модулем взвешивания"
                Case 24
                    _message = "Установлена нагрузка на платформу при включении весового устройства"
                Case 25
                    _message = "Весовое устройство неисправно"
                Case Else
                    _message = "Неописанная в документации ошибка"
            End Select
            Return _message

        End Get
    End Property

    Public ReadOnly Property LastResponseNum As Long Implements IScaleDriver.LastResponseNum
        Get
            Return _response
        End Get
    End Property

    Public ReadOnly Property isConnected As Boolean Implements IScaleDriver.isConnected
        Get
            Select Case _response
                Case 1, 2, 3, 23
                    Return False
                Case Else
                    Return True
            End Select
        End Get
    End Property

    Public ReadOnly Property isScaleAlarm As Boolean Implements IScaleDriver.isScaleAlarm
        Get
            Select Case _response
                Case 0, 1, 2, 3, 23
                    Return False
                Case Else
                    Return True
            End Select
        End Get
    End Property

    Public Sub OpenConnection() Implements IScaleDriver.OpenConnection
        Try
            _response = _scale.OpenConnection()
        Catch ex As Exception
            Throw New InvalidOperationException($"Ошибка драйвера OpenConnection(): {LastResponseText}", ex)
        End Try

    End Sub

    Public Sub CloseConnection() Implements IScaleDriver.CloseConnection
        Try
            _response = _scale.CloseConnection()
        Catch ex As Exception
            Throw New InvalidOperationException($"Ошибка драйвера CloseConnection(): {LastResponseText}", ex)
        End Try

    End Sub

    Public Sub SetToZero() Implements IScaleDriver.SetToZero
        Try
            _response = _scale.SetToZero()
        Catch ex As Exception
            Throw New InvalidOperationException($"Ошибка драйвера SetToZero(): {LastResponseText}", ex)
        End Try

    End Sub

    Public Sub ReadWeight() Implements IScaleDriver.ReadWeight
        Try
            _response = _scale.ReadWeight()
        Catch ex As Exception
            Throw New InvalidOperationException($"Ошибка драйвера ReadWeight(): {LastResponseText}", ex)
        End Try

    End Sub
End Class

File: Scalemon.MassaKInterop/ScaleDriver100.cs
using System;
using Microsoft.VisualBasic.CompilerServices;

namespace Scalemon.MassaKInterop
{

    public class ScaleDriver100 : IScaleDriver
    {

        private MassaKDriver100.Scales _scale;
        private long _response;


        public ScaleDriver100()
        {
            _scale = new MassaKDriver100.Scales();
            _response = 0L;
        }

        public string PortConnection
        {
            get
            {
                return _scale.Connection;
            }
            set
            {
                _scale.Connection = value;
            }
        }

        public decimal Weight
        {
            get
            {
                return _scale.Weight / 100m;
            }
        }

        public bool Stable
        {
            get
            {
                return _scale.Stable == 1;
            }
        }

        public string LastResponseText
        {
            get
            {
                string _message;
                switch (_response)
                {
                    case 0L:
                        {
                            _message = "Ошибок нет";
                            break;
                        }

                    case 1L:
                        {
                            _message = "Связь с весами не установлена";
                            break;
                        }

                    case 2L:
                        {
                            _message = "Ошибка обмена данных с весами";
                            break;
                        }

                    case 3L:
                        {
                            _message = "Весы не готовы к передаче данных";
                            break;
                        }

                    case 4L:
                        {
                            _message = "Параметр не поддерживается весами";
                            break;
                        }

                    case 5L:
                        {
                            _message = "Установка параметра невозможна";
                            break;
                        }

                    case 7L:
                        {
                            _message = "Невозможно выполнить команду или команда не поддерживается";
                            break;
                        }

                    case 8L:
                        {
                            _message = "Нагрузка на весовом устройстве превышает НПВ";
                            break;
                        }

                    case 9L:
                        {
                            _message = "Весовое устройство не в режиме взвешивания";
                            break;
                        }

                    case 10L:
                        {
                            _message = "Ошибка входных данных";
                            break;
                        }

                    case 11L:
                        {
                            _message = "Ошибка сохранения данных";
                            break;
                        }

                    case 16L:
                        {
                            _message = "Интерфейс Wi-Fi не поддерживается";
                            break;
                        }

                    case 17L:
                        {
                            _message = "Интерфейс Ethernet не поддерживается";
                            break;
                        }

                    case 21L:
                        {
                            _message = "Установка нуля невозможна из-за наличия нагрузки на платформе";
                            break;
                        }

                    case 23L:
                        {
                            _message = "Нет связи с модулем взвешивания";
                            break;
                        }

                    case 24L:
                        {
                            _message = "Установлена нагрузка на платформу при включении весового устройства";
                            break;
                        }

                    case 25L:
                        {
                            _message = "Весовое устройство неисправно";
                            break;
                        }

                    default:
                        {
                            _message = "Неописанная в документации ошибка";
                            break;
                        }
                }
                return _message;

            }
        }

        public long LastResponseNum
        {
            get
            {
                return _response;
            }
        }

        public bool isConnected
        {
            get
            {
                switch (_response)
                {
                    case 1L:
                    case 2L:
                    case 3L:
                    case 23L:
                        {
                            return false;
                        }

                    default:
                        {
                            return true;
                        }
                }
            }
        }

        public bool isScaleAlarm
        {
            get
            {
                switch (_response)
                {
                    case 0L:
                    case 1L:
                    case 2L:
                    case 3L:
                    case 23L:
                        {
                            return false;
                        }

                    default:
                        {
                            return true;
                        }
                }
            }
        }

        public void OpenConnection()
        {
            try
            {
                _response = _scale.OpenConnection();
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException($"Ошибка драйвера OpenConnection(): {LastResponseText}", ex);
            }

        }

        public void CloseConnection()
        {
            try
            {
                _response = _scale.CloseConnection();
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException($"Ошибка драйвера CloseConnection(): {LastResponseText}", ex);
            }

        }

        public void SetToZero()
        {
            try
            {
                var result = _scale.SetZero();
                _response = Convert.ToInt64(result);
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException($"Ошибка драйвера SetToZero(): {LastResponseText}", ex);
            }

        }

        public void ReadWeight()
        {
            try
            {
                _response = _scale.ReadWeight();
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException($"Ошибка драйвера ReadWeight(): {LastResponseText}", ex);
            }

        }
    }
}